File: examples/result_contract/mcp-typescript-sdk-v2/generate-evidence.mjs
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import {
  Client,
  StreamableHTTPClientTransport,
} from "@modelcontextprotocol/client";
import {
  McpServer,
  createMcpHandler,
  fromJsonSchema,
} from "@modelcontextprotocol/server";
import { z } from "zod";

const outputArgument = process.argv[2];
if (!outputArgument) {
  throw new Error("Usage: pnpm run evidence -- <output-directory>");
}

const exampleDirectory = dirname(fileURLToPath(import.meta.url));
const repositoryRoot = resolve(exampleDirectory, "../../..");
const schemaPath = resolve(
  repositoryRoot,
  "src/qzx/resources/schemas/result-contract-v1.schema.json",
);
const outputDirectory = resolve(outputArgument);
const contractSchema = JSON.parse(await readFile(schemaPath, "utf8"));

function buildServer() {
  const server = new McpServer({
    name: "qzx-result-contract-sdk-evidence",
    version: "1.0.0",
  });

  server.registerTool(
    "lookup_widget",
    {
      description:
        "Look up one synthetic widget for a QZX interoperability test.",
      inputSchema: z.object({ fail: z.boolean() }),
      outputSchema: fromJsonSchema(contractSchema),
    },
    async ({ fail }) => {
      const result = fail
        ? {
            success: false,
            message: "The requested widget was not found.",
            error_code: "widget_not_found",
            details: { widget_id: "missing-widget" },
          }
        : {
            success: true,
            message: "The requested widget was returned.",
            details: { widget_id: "widget-1", status: "ready" },
          };

      return {
        content: [{ type: "text", text: JSON.stringify(result) }],
        structuredContent: result,
        isError: fail,
      };
    },
  );
  return server;
}

async function parseWireDocuments(capture) {
  const body = await capture.response;
  if (capture.contentType.includes("application/json")) {
    return [JSON.parse(body)];
  }
  return body
    .split(/\r?\n\r?\n/)
    .map((event) =>
      event
        .split(/\r?\n/)
        .filter((line) => line.startsWith("data:"))
        .map((line) => line.slice("data:".length).trimStart())
        .join("\n"),
    )
    .filter(Boolean)
    .map((data) => JSON.parse(data));
}

const handler = createMcpHandler(buildServer);
const wireCaptures = [];
const transport = new StreamableHTTPClientTransport(
  new URL("http://test.local/mcp"),
  {
    fetch: async (url, init) => {
      const request = new Request(url, init);
      const requestDocument =
        request.method === "POST" ? await request.clone().json() : null;
      const response = await handler.fetch(request);
      if (requestDocument?.method === "tools/call") {
        wireCaptures.push({
          request: requestDocument,
          response: response.clone().text(),
          contentType: response.headers.get("content-type") ?? "",
        });
      }
      return response;
    },
  },
);

const client = new Client(
  { name: "qzx-result-contract-sdk-evidence-client", version: "1.0.0" },
  { versionNegotiation: { mode: { pin: "2026-07-28" } } },
);

try {
  await client.connect(transport);
  if (client.getProtocolEra() !== "modern") {
    throw new Error(
      `Expected a modern MCP connection, got ${client.getProtocolEra()}.`,
    );
  }

  const listedTools = await client.listTools();
  const toolDefinition = listedTools.tools.find(
    (tool) => tool.name === "lookup_widget",
  );
  if (!toolDefinition) {
    throw new Error("The official MCP client did not discover lookup_widget.");
  }

  const success = await client.callTool({
    name: "lookup_widget",
    arguments: { fail: false },
  });
  const failure = await client.callTool({
    name: "lookup_widget",
    arguments: { fail: true },
  });
  if (
    success.isError === true ||
    success.structuredContent?.success !== true ||
    failure.isError !== true ||
    failure.structuredContent?.success !== false
  ) {
    throw new Error(
      "The official MCP client observed an inconsistent result pair.",
    );
  }

  const completedResults = new Map();
  for (const capture of wireCaptures) {
    const documents = await parseWireDocuments(capture);
    const response = documents.find(
      (document) => document.id === capture.request.id,
    );
    if (!response?.result) {
      throw new Error(
        `No JSON-RPC tool result was captured for request ${capture.request.id}.`,
      );
    }
    completedResults.set(
      Boolean(capture.request.params.arguments.fail),
      response.result,
    );
  }

  const rawSuccess = completedResults.get(false);
  const rawFailure = completedResults.get(true);
  if (
    rawSuccess?.resultType !== "complete" ||
    rawFailure?.resultType !== "complete"
  ) {
    throw new Error("The complete MCP 2026-07-28 wire pair was not captured.");
  }

  await mkdir(outputDirectory, { recursive: true });
  const evidenceMetadata = {
    success: true,
    message: "Official MCP TypeScript SDK v2 reference evidence generated.",
    details: {
      evidence_kind: "qzx_maintained_reference",
      independent_adoption: false,
      protocol: "2026-07-28",
      protocol_era: client.getProtocolEra(),
      transport: "in_process_streamable_http_fetch",
      packages: {
        "@modelcontextprotocol/client": "2.0.0",
        "@modelcontextprotocol/server": "2.0.0",
        zod: "4.4.3",
      },
      runtime: {
        node: process.version,
        platform: process.platform,
        architecture: process.arch,
      },
      output_directory: outputDirectory,
      files: [
        "tool-definition.json",
        "success.json",
        "failure.json",
        "evidence-metadata.json",
      ],
    },
  };
  for (const [filename, document] of [
    ["tool-definition.json", toolDefinition],
    ["success.json", rawSuccess],
    ["failure.json", rawFailure],
    ["evidence-metadata.json", evidenceMetadata],
  ]) {
    await writeFile(
      resolve(outputDirectory, filename),
      `${JSON.stringify(document, null, 2)}\n`,
      "utf8",
    );
  }

  process.stdout.write(`${JSON.stringify(evidenceMetadata, null, 2)}\n`);
} finally {
  await client.close();
  await handler.close();
}

File: examples/result_contract/mcp-python-sdk-v2/generate_evidence.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Generate QZX Result Contract evidence with the official MCP Python SDK v2."""

from __future__ import annotations

import asyncio
from importlib.metadata import version
import json
from pathlib import Path
import platform
import sys
from typing import Any

from mcp.client import Client
from mcp.server import Server
import mcp.types as types


PROTOCOL_VERSION = "2026-07-28"
TOOL_NAME = "lookup_widget"
EXAMPLE_DIRECTORY = Path(__file__).resolve().parent
REPOSITORY_ROOT = EXAMPLE_DIRECTORY.parents[2]
SCHEMA_PATH = (
    REPOSITORY_ROOT
    / "src"
    / "qzx"
    / "resources"
    / "schemas"
    / "result-contract-v1.schema.json"
)


def parse_output_directory(arguments: list[str]) -> Path:
    """Accept one output directory and tolerate a package-runner separator."""

    positional = [argument for argument in arguments if argument != "--"]
    if len(positional) != 1:
        raise ValueError("Usage: python generate_evidence.py <output-directory>")
    return Path(positional[0]).expanduser().resolve()


def success_document() -> dict[str, Any]:
    return {
        "success": True,
        "message": "The requested widget was returned.",
        "details": {"widget_id": "widget-1", "status": "ready"},
    }


def failure_document() -> dict[str, Any]:
    return {
        "success": False,
        "message": "The requested widget was not found.",
        "error_code": "widget_not_found",
        "details": {"widget_id": "missing-widget"},
    }


def build_server(contract_schema: dict[str, Any]) -> Server[Any]:
    tool = types.Tool(
        name=TOOL_NAME,
        description="Look up one synthetic widget for a QZX interoperability test.",
        input_schema={
            "type": "object",
            "properties": {"fail": {"type": "boolean"}},
            "required": ["fail"],
            "additionalProperties": False,
        },
        output_schema=contract_schema,
    )

    async def list_tools(
        _context: Any,
        _params: types.PaginatedRequestParams | None,
    ) -> types.ListToolsResult:
        return types.ListToolsResult(tools=[tool])

    async def call_tool(
        _context: Any,
        params: types.CallToolRequestParams,
    ) -> types.CallToolResult:
        if params.name != TOOL_NAME:
            raise ValueError(f"Unknown synthetic tool: {params.name}")
        fail = bool((params.arguments or {}).get("fail"))
        document = failure_document() if fail else success_document()
        return types.CallToolResult(
            content=[
                types.TextContent(
                    text=json.dumps(
                        document,
                        ensure_ascii=False,
                        separators=(",", ":"),
                        sort_keys=True,
                    )
                )
            ],
            structured_content=document,
            is_error=fail,
        )

    return Server(
        "qzx-result-contract-python-sdk-evidence",
        version="1.0.0",
        on_list_tools=list_tools,
        on_call_tool=call_tool,
    )


def wire_document(model: Any) -> dict[str, Any]:
    """Serialize an SDK model with its MCP wire aliases."""

    return model.model_dump(by_alias=True, exclude_none=True, mode="json")


async def generate(output_directory: Path) -> dict[str, Any]:
    contract_schema = json.loads(SCHEMA_PATH.read_text(encoding="utf-8"))
    server = build_server(contract_schema)

    async with Client(server, mode=PROTOCOL_VERSION) as client:
        if client.protocol_version != PROTOCOL_VERSION:
            raise RuntimeError(
                f"Expected MCP {PROTOCOL_VERSION}, got {client.protocol_version}."
            )

        listed_tools = await client.list_tools()
        matching_tools = [tool for tool in listed_tools.tools if tool.name == TOOL_NAME]
        if len(matching_tools) != 1:
            raise RuntimeError(
                "The official MCP client did not discover lookup_widget."
            )

        success_result = await client.call_tool(TOOL_NAME, {"fail": False})
        failure_result = await client.call_tool(TOOL_NAME, {"fail": True})

    tool_definition = wire_document(matching_tools[0])
    success = wire_document(success_result)
    failure = wire_document(failure_result)

    if tool_definition.get("outputSchema") != contract_schema:
        raise RuntimeError(
            "The official SDK changed the canonical inline output schema."
        )
    if (
        success.get("resultType") != "complete"
        or failure.get("resultType") != "complete"
    ):
        raise RuntimeError("The official SDK did not retain MCP completed-result tags.")
    if success.get("isError") is not False or failure.get("isError") is not True:
        raise RuntimeError("The official SDK observed an inconsistent result pair.")

    evidence_metadata = {
        "evidence_kind": "qzx_maintained_reference",
        "independent_adoption": False,
        "protocol": PROTOCOL_VERSION,
        "protocol_era": "modern",
        "transport": "in_process_direct_dispatcher",
        "jsonrpc_framing_exercised": False,
        "serialization": "pydantic_model_dump_by_alias",
        "packages": {
            "mcp": version("mcp"),
            "mcp-types": version("mcp-types"),
            "pydantic": version("pydantic"),
        },
        "runtime": {
            "implementation": platform.python_implementation(),
            "python": platform.python_version(),
            "platform": sys.platform,
            "architecture": platform.machine(),
        },
        "output_directory": str(output_directory),
        "files": [
            "tool-definition.json",
            "success.json",
            "failure.json",
            "evidence-metadata.json",
        ],
    }

    output_directory.mkdir(parents=True, exist_ok=True)
    documents = (
        ("tool-definition.json", tool_definition),
        ("success.json", success),
        ("failure.json", failure),
        ("evidence-metadata.json", evidence_metadata),
    )
    for filename, document in documents:
        (output_directory / filename).write_text(
            json.dumps(document, ensure_ascii=False, indent=2, sort_keys=True) + "\n",
            encoding="utf-8",
            newline="\n",
        )
    return evidence_metadata


def main() -> int:
    output_directory = parse_output_directory(sys.argv[1:])
    metadata = asyncio.run(generate(output_directory))
    print(
        json.dumps(
            {
                "success": True,
                "message": "Official MCP Python SDK v2 reference evidence generated.",
                "details": metadata,
            },
            ensure_ascii=False,
            indent=2,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
